README.md
[![Docker](https://img.shields.io/docker/v/xchikux/backend?color=0db7ed&label=docker&sort=date)](https://hub.docker.com/r/xchikux/trampoline)

<br /><br />

<p align="center"><img align="center" src="https://psync.club/web/psync.jpeg" height="96" alt="Trampoline" /></p>
<h1 align="center">Trampoline</h1>
<h3 align="center">Elevate yourself. Land your email.</h3>
<h4 align="center">Comes with a <a href="./backend">⚙️ HTTP backend</a>.</h4>

<br /><br /><br />

## 👉 Live Demo: https://reacher.email

<img src="https://psync.club/web/psync.jpeg" height="68" align="left" />

This is open-source, but I also offer a **SaaS** solution that has `trampoline` packaged in a nice friendly web interface. If you are interested, find out more at [Trampoline](https://psync.club/?ref=github). If you have any questions, you can contact me at srikanth@psync.dev

<br />

## Get Started

3 non-SaaS ways to get started with `trampoline`.

### 1. ⚙️ HTTP backend using Docker (popular method 🥇) [[Full docs](./backend/README.md)]

This option allows you to run a HTTP backend using Docker 🐳, on a cloud instance or your own server. Please note that outbound port 25 must be open.

```bash
docker run <coming_soon>
```

Then send a `POST http://localhost:8080/v0/check_email` request with the following body:

```js
{
    "to_email": "someone@gmail.com",
    "from_email": "my@my-server.com", // (optional) email to use in the `FROM` SMTP command, defaults to "user@example.org"
    "hello_name": "my-server.com",    // (optional) name to use in the `EHLO` SMTP command, defaults to "localhost"
    "proxy": {                        // (optional) SOCK5 proxy to run the verification through, default is empty
        "host": "my-proxy.io",
        "port": 1080,
        "username": "me",             // (optional) Proxy username
        "password": "pass"            // (optional) Proxy password
    },
    "smtp_port": 587                  // (optional) SMTP port to do the email verification, defaults to 25
}
```

### 2. Download the CLI [[Full docs](./cli/README.md)]

> Note: The CLI binary doesn't connect to any backend, it checks the email directly from your computer.

Head to the [releases page](https://github.com/reacherhq/check-if-email-exists/releases) and download the binary for your platform.

```bash
> $ trampoline --help
trampoline 0.9.2
Check if an email address exists without sending an email.

USAGE:
    trampoline [FLAGS] [OPTIONS] [TO_EMAIL]
```

Check out the [dedicated README.md](./cli/README.md) for all options and flags.

### 3. Programmatic Usage [[Full docs](https://docs.rs/check-if-email-exists)]

In your own Rust project, you can add `trampoline` in your `Cargo.toml`:

```toml
[dependencies]
check-if-email-exists = "0.9"
```

And use it in your code as follows:

```rust
use trampoline::{check_email, CheckEmailInput, CheckEmailInputProxy};

async fn check() {
    // Let's say we want to test the deliverability of someone@gmail.com.
    let mut input = CheckEmailInput::new(vec!["someone@gmail.com".into()]);

    // Verify this email, using async/await syntax.
    let result = check_email(&input).await;

    // `result` is a `Vec<CheckEmailOutput>`, where the CheckEmailOutput
    // struct contains all information about our email.
    println!("{:?}", result);
}
```

The reference docs are hosted on [docs.rs](https://docs.rs/check-if-email-exists).

## ✈️ JSON Output

The output will be a JSON with the below format, the fields should be self-explanatory. For `someone@gmail.com` (note that it is disabled by Gmail), here's the exact output:

```json
{
  "input": "someone@gmail.com",
  "is_reachable": "invalid",
  "misc": {
    "is_disposable": false,
    "is_role_account": false
  },
  "mx": {
    "accepts_mail": true,
    "records": [
      "alt3.gmail-smtp-in.l.google.com.",
      "gmail-smtp-in.l.google.com.",
      "alt1.gmail-smtp-in.l.google.com.",
      "alt4.gmail-smtp-in.l.google.com.",
      "alt2.gmail-smtp-in.l.google.com."
    ]
  },
  "smtp": {
    "can_connect_smtp": true,
    "has_full_inbox": false,
    "is_catch_all": false,
    "is_deliverable": false,
    "is_disabled": true
  },
  "syntax": {
    "domain": "gmail.com",
    "is_valid_syntax": true,
    "username": "someone",
    "suggestion": null
  }
}
```

## What Does This Tool Check?

| Included? | Feature                                       | Description                                                                                                                     | JSON field                                                                  |
| --------- | --------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------- | --------------------------------------------------------------------------- |
| ✅        | **Email reachability**                        | How confident are we in sending an email to this address? Can be one of `safe`, `risky`, `invalid` or `unknown`.                | `is_reachable`                                                              |
| ✅        | **Syntax validation**                         | Is the address syntactically valid?                                                                                             | `syntax.is_valid_syntax`                                                    |
| ✅        | **DNS records validation**                    | Does the domain of the email address have valid MX DNS records?                                                                 | `mx.accepts_mail`                                                           |
| ✅        | **Disposable email address (DEA) validation** | Is the address provided by a known [disposable email address](https://en.wikipedia.org/wiki/Disposable_email_address) provider? | `misc.is_disposable`                                                        |
| ✅        | **SMTP server validation**                    | Can the mail exchanger of the email address domain be contacted successfully?                                                   | `smtp.can_connect_smtp`                                                     |
| ✅        | **Email deliverability**                      | Is an email sent to this address deliverable?                                                                                   | `smtp.is_deliverable`                                                       |
| ✅        | **Mailbox disabled**                          | Has this email address been disabled by the email provider?                                                                     | `smtp.is_disabled`                                                          |
| ✅        | **Full inbox**                                | Is the inbox of this mailbox full?                                                                                              | `smtp.has_full_inbox`                                                       |
| ✅        | **Catch-all address**                         | Is this email address a [catch-all](https://debounce.io/blog/help/what-is-a-catch-all-or-accept-all/) address?                  | `smtp.is_catch_all`                                                         |
| ✅        | **Role account validation**                   | Is the email address a well-known role account?                                                                                 | `misc.is_role_account`                                                      |
| ✅        | **Gravatar Url**                              | The url of the [Gravatar](https://gravatar.com/) email address profile picture                                                  | `misc.gravatar_url`                                                         |
| 🔜        | **Free email provider check**                 | Is the email address bound to a known free email provider?                                                                      | [Issue #89](https://github.com/reacherhq/check-if-email-exists/issues/89)   |
| 🔜        | **Syntax validation, provider-specific**      | According to the syntactic rules of the target mail provider, is the address syntactically valid?                               | [Issue #90](https://github.com/reacherhq/check-if-email-exists/issues/90)   |
| 🔜        | **Honeypot detection**                        | Does email address under test hide a [honeypot](https://en.wikipedia.org/wiki/Spamtrap)?                                        | [Issue #91](https://github.com/reacherhq/check-if-email-exists/issues/91)   |
| 🔜        | **Have I Been Pwned?**                        | Has this email been compromised in a [data breach](https://haveibeenpwned.com/)?                                                | [Issue #289](https://github.com/reacherhq/check-if-email-exists/issues/289) |

## 🤔 Why?

Many online services (https://hunter.io, https://verify-email.org, https://email-checker.net) offer this service for a paid fee. Here is an open-source alternative to those tools.

### Open source license

If you are creating an open-source application under a license compatible with the GNU Affero GPL license v3, you may use `trampoline` under the terms of the [AGPL-3.0](./LICENSE.AGPL).

## ❓ FAQ

#### What does `is_reachable: "unknown"` mean?

This means that the server does not allow real-time verification of an email right now. It may happen for multiple reasons: your IP is blacklisted, the SMTP port 25 is blocked, the email account is momentarily receiving too many emails (spam protection)... or the email provider simply does not allow real-time verification at all. The details of this `"unknown"` case can be found in the `smtp.error` and `mx.error` fields.

#### The library hangs/takes a long time/doesn't show anything after 1 minute.

Most ISPs block outgoing SMTP requests through port 25, to prevent spam. `trampoline` needs to have this port open to make a connection to the email's SMTP server, so won't work behind these ISPs, and will instead hang until it times out. There's unfortunately no easy workaround for this problem, see for example [this StackOverflow thread](https://stackoverflow.com/questions/18139102/how-to-get-around-an-isp-block-on-port-25-for-smtp). One solution is to rent a Linux cloud server with a static IP and no blocked ports, see for example our [Self-Host Guide](https://help.reacher.email/self-host-guide) for which cloud providers open port 25.

To see in detail what the binary is doing behind the scenes, run it in verbose mode with `RUST_LOG=debug` to see the logs.

#### I have another question.

Feel free to check out Reacher's [FAQ](https://help.reacher.email/faq).

## 🔨 Build From Source

Build the [CLI from source](./cli/README.md#build-from-source)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [0.9.0](https://github.com/xchikux/trampoline/compare/v0.8.32...v0.9.0) (2022-08-15)

### ⚠ BREAKING CHANGES

- The `RUST_LOG` target has been changed from `trampoline` to `trampoline`.

```diff
- RUST_LOG=trampoline=debug cargo run
- RUST_LOG=trampoline=debug cargo run
```

- The library's main function `check_email`'s argument `CheckEmailInput` nows takes a single `to_email` field, instead of a `to_emails: Vec<String>`

```diff
pub struct CheckEmailInput {
- pub to_emails: Vec<String>,
+ pub to_email: String,
  // --snip--
}
```

- The `--http` flag has been removed from `trampoline` CLI. To run a HTTP backend, please refer to the [backend](./backend) subfolder.

### Features

- Move `backend` code to this repo ([#1138](https://github.com/xchikux/trampoline/issues/1138)) ([0dc6053](https://github.com/xchikux/trampoline/commit/0dc60531d26efb217137347ef2b6aaf678d94238))
- Revert back to `check_email` input with single email ([#1150](https://github.com/xchikux/trampoline/issues/1150)) ([ce1ba53](https://github.com/xchikux/trampoline/commit/ce1ba5346849b578a0ed30b1d72096f15cfbc09d)), closes [#65](https://github.com/xchikux/trampoline/issues/65)

- Change RUST_LOG target to `trampoline` ([#1152](https://github.com/xchikux/trampoline/issues/1152)) ([7e87be2](https://github.com/xchikux/trampoline/commit/7e87be26f1e35a6936bfc967c872cd42b93fd256))
- Remove HTTP backend from CLI ([#1151](https://github.com/xchikux/trampoline/issues/1151)) ([7184372](https://github.com/xchikux/trampoline/commit/71843720c9b87fa0e43fa482a35ef074435bf562))

### [0.8.32](https://github.com/xchikux/trampoline/compare/v0.8.31...v0.8.32) (2022-08-13)

### Features

- Use opportunistic STARTTLS by default ([#1079](https://github.com/xchikux/trampoline/issues/1079)) ([54911f0](https://github.com/xchikux/trampoline/commit/54911f0a8ec51e753f757878021e933609cff868))

### Bug Fixes

- Fix parsing some invalid emails ([cb65c0f](https://github.com/xchikux/trampoline/commit/cb65c0f4767b2f163f48054652f7652b6d0b6043))
- Syntax also check using using `mailchecker` ([8385bec](https://github.com/xchikux/trampoline/commit/8385bec6fedc0912881800442bffda5b33c2f394))

### [0.8.31](https://github.com/xchikux/trampoline/compare/v0.8.30...v0.8.31) (2022-08-10)

### Improvements

- Bump packages incl. `async-smtp` ([#1136](https://github.com/amaurymartiny/trampoline/issues/1136)) ([43b47ea](https://github.com/amaurymartiny/trampoline/commit/cbdab7c35bc4ce7f7da79fa9316727cc840da4e4))

### [0.8.30](https://github.com/amaurymartiny/trampoline/compare/v0.8.29...v0.8.30) (2022-06-02)

### Features

- Add `smtp.error.description` field for human-readable description of error ([#1111](https://github.com/amaurymartiny/trampoline/issues/1111)) ([43b47ea](https://github.com/amaurymartiny/trampoline/commit/43b47ea2b9250f2c6d58c8a0ec4340066169c169))

### Bug Fixes

- Fix `has_full_inbox` check too lenient ([93de444](https://github.com/amaurymartiny/trampoline/commit/93de444dfa7c6d66061570115be8f53f0647c431))

### [0.8.29](https://github.com/amaurymartiny/trampoline/compare/v0.8.28...v0.8.29) (2022-03-02)

### Features

- Loop through all MX servers ([#1070](https://github.com/amaurymartiny/trampoline/issues/1070)) ([11e6a06](https://github.com/amaurymartiny/trampoline/commit/11e6a06a67f5893b729c76d1a33667f83d63c836))

### [0.8.28](https://github.com/amaurymartiny/trampoline/compare/v0.8.27...v0.8.28) (2022-02-11)

### Features

- Add proxy username/password ([#1057](https://github.com/amaurymartiny/trampoline/issues/1057)) ([d9583c6](https://github.com/amaurymartiny/trampoline/commit/d9583c6ae0d3353a5135dd157999cf579b308d6d))

### [0.8.27](https://github.com/amaurymartiny/trampoline/compare/v0.8.26...v0.8.27) (2022-02-07)

### Features

- Allow user to define SMTP client security for TLS ([#1043](https://github.com/amaurymartiny/trampoline/issues/1043)) ([bc722ff](https://github.com/amaurymartiny/trampoline/commit/bc722ff1a9b30747308a3b3b5959d73e5e853292))
- Break SmtpError into `{Helo,Connect,ConnectWithStream,MailFrom,RcptTo,Close}Error` ([#1055](https://github.com/amaurymartiny/trampoline/issues/1055)) ([64e5193](https://github.com/amaurymartiny/trampoline/commit/64e5193c48a6bf4c080e79daeefd1c98dadffd5d))

### [0.8.26](https://github.com/amaurymartiny/trampoline/compare/v0.8.25...v0.8.26) (2022-01-26)

### Features

- Add SMTP retries to avoid greylisting ([#1041](https://github.com/amaurymartiny/trampoline/issues/1041)) ([b451a1e](https://github.com/amaurymartiny/trampoline/commit/b451a1e93a6ccf025c78d56dee7439ad607c8507))

### Bug Fixes

- Use std::default for deriving ([#1015](https://github.com/amaurymartiny/trampoline/issues/1015)) ([03720f0](https://github.com/amaurymartiny/trampoline/commit/03720f027fd68d5ea5ae538aa567a621f4a65fe3))

### [0.8.25](https://github.com/amaurymartiny/trampoline/compare/v0.8.24...v0.8.25) (2021-10-05)

### Features

- Add possibility to set SMTP port ([#985](https://github.com/amaurymartiny/trampoline/issues/985)) ([cdabdf8](https://github.com/amaurymartiny/trampoline/commit/cdabdf80e858908d6c33e1273dfdc1fef0f78d35))

### Bug Fixes

- Use async_std_resolver::resolver_from_system_conf ([#982](https://github.com/amaurymartiny/trampoline/issues/982)) ([376c3b0](https://github.com/amaurymartiny/trampoline/commit/376c3b0d4743ccc60a1df2a9fa3e9f2f5cd68178))
- Use TLS when available ([#964](https://github.com/amaurymartiny/trampoline/issues/964)) ([aed11d2](https://github.com/amaurymartiny/trampoline/commit/aed11d2e15b6b7688ecaf856824ca6effbb5d21b))

### [0.8.24](https://github.com/amaurymartiny/trampoline/compare/v0.8.23...v0.8.24) (2021-07-03)

### Features


core/src/util/input_output.rs
// check-if-email-exists
// Copyright (C) 2018-2022 Reacher

// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published
// by the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.

// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

use std::time::Duration;

use async_smtp::{ClientSecurity, ClientTlsParameters};
use serde::{ser::SerializeMap, Deserialize, Serialize, Serializer};

use crate::misc::{MiscDetails, MiscError};
use crate::mx::{MxDetails, MxError};
use crate::smtp::{SmtpDetails, SmtpError, SmtpErrorDesc};
use crate::syntax::SyntaxDetails;

/// Perform the email verification via a specified proxy. The usage of a proxy
/// is optional.
#[derive(Debug, Default, Clone, Deserialize, Serialize)]
pub struct CheckEmailInputProxy {
	/// Use the specified SOCKS5 proxy host to perform email verification.
	pub host: String,
	/// Use the specified SOCKS5 proxy port to perform email verification.
	pub port: u16,
	/// Username to pass to proxy authentication.
	pub username: Option<String>,
	/// Password to pass to proxy authentication.
	pub password: Option<String>,
}

/// Define how to apply TLS to a SMTP client connection. Will be converted into
/// async_smtp::ClientSecurity.
#[derive(Debug, Clone, Copy, Deserialize, Serialize)]
pub enum SmtpSecurity {
	/// Insecure connection only (for testing purposes).
	None,
	/// Start with insecure connection and use `STARTTLS` when available.
	Opportunistic,
	/// Start with insecure connection and require `STARTTLS`.
	Required,
	/// Use TLS wrapped connection.
	Wrapper,
}

impl Default for SmtpSecurity {
	fn default() -> Self {
		Self::Opportunistic
	}
}

impl SmtpSecurity {
	pub fn to_client_security(self, tls_params: ClientTlsParameters) -> ClientSecurity {
		match self {
			Self::None => ClientSecurity::None,
			Self::Opportunistic => ClientSecurity::Opportunistic(tls_params),
			Self::Required => ClientSecurity::Required(tls_params),
			Self::Wrapper => ClientSecurity::Wrapper(tls_params),
		}
	}
}

/// Builder pattern for the input argument into the main `email_exists`
/// function.
#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(default)]
pub struct CheckEmailInput {
	/// The email to validate.
	pub to_email: String,
	/// Email to use in the `MAIL FROM:` SMTP command.
	///
	/// Defaults to "vegetoyesh@gmail.com", which is an unused addressed
	/// owned by Reacher.
	pub from_email: String,
	/// Name to use in the `EHLO:` SMTP command.
	///
	/// Defaults to "gmail.com" (note: "localhost" is not a FQDN).
	pub hello_name: String,
	/// Perform the email verification via the specified SOCK5 proxy. The usage of a
	/// proxy is optional.
	pub proxy: Option<CheckEmailInputProxy>,
	/// SMTP port to use for email validation. Generally, ports 25, 465, 587
	/// and 2525 are used.
	///
	/// Defaults to 25.
	pub smtp_port: u16,
	/// Add timeout for the SMTP verification step. Set to None if you don't
	/// want to use a timeout.
	///
	/// Defaults to 12s (more than 10s, but when run twice less than 30s).
	pub smtp_timeout: Option<Duration>,
	/// For Yahoo email addresses, use Yahoo's API instead of connecting
	/// directly to their SMTP servers.
	///
	/// Defaults to true.
	pub yahoo_use_api: bool,
	/// For Gmail email addresses, use Gmail's API instead of connecting
	/// directly to their SMTP servers.
	///
	/// Defaults to false.
	pub gmail_use_api: bool,
	/// For Microsoft 365 email addresses, use OneDrive's API instead of
	/// connecting directly to their SMTP servers.
	///
	/// Defaults to false.
	pub microsoft365_use_api: bool,
	// Whether to check if a gravatar image is existing for the given email.
	//
	// Defaults to false.
	pub check_gravatar: bool,
	/// Check if a the email address is present in HaveIBeenPwned API.
	// If the api_key is filled, HaveIBeenPwned API is checked
	pub haveibeenpwned_api_key: Option<String>,
	/// For Hotmail/Outlook email addresses, use a headless navigator
	/// connecting to the password recovery page instead of the SMTP server.
	/// This assumes you have a WebDriver compatible process running, then pass
	/// its endpoint, usually http://localhost:4444. We recommend running
	/// chromedriver (and not geckodriver) as it allows parallel requests.
	///
	/// Defaults to None.
	#[cfg(feature = "headless")]
	pub hotmail_use_headless: Option<String>,
	/// Number of retries of SMTP connections to do.
	///
	/// Defaults to 2 to avoid greylisting.
	pub retries: usize,
	/// How to apply TLS to a SMTP client connection.
	///
	/// Defaults to Opportunistic.
	pub smtp_security: SmtpSecurity,
	/// **IMPORTANT:** This is a beta feature, and might be completely removed,
	/// or moved somewhere else, before the next release.
	///
	/// List of domains to skip when doing an SMTP connection, because we know
	/// they return "unknown". For each string in this list, we check if the MX
	/// record **contains** the string; if yes, we return an error saying the
	/// SMTP verification is skipped.
	///
	/// Related issue: https://github.com/reacherhq/check-if-email-exists/issues/937
	///
	/// ## Example
	///
	/// If you want to skip Zoho emails, it's good enough to ".zoho.com." to
	/// the list. This way it will skip all Zoho emails as their MX records
	/// show "mx{N}.zoho.com.". Simply putting "zoho" might give false
	/// positives if you have an email provider like "mycustomzoho.com".
	///
	/// Defaults to: [""]
	pub skipped_domains: Vec<String>,
}

impl Default for CheckEmailInput {
	fn default() -> Self {
		CheckEmailInput {
			to_email: "".into(),
			from_email: "vegetoyesh@gmail.com".into(), // Unused, owned by Reacher
			hello_name: "gmail.com".into(),
			#[cfg(feature = "headless")]
			hotmail_use_headless: None,
			proxy: None,
			smtp_port: 25,
			smtp_security: SmtpSecurity::default(),
			smtp_timeout: Some(Duration::from_secs(12)),
			yahoo_use_api: true,
			gmail_use_api: false,
			microsoft365_use_api: false,
			check_gravatar: false,
			haveibeenpwned_api_key: None,
			retries: 2,
			skipped_domains: vec![
				// on @bluewin.ch
				// - mx-v02.bluewin.ch.
				".bluewin.ch.".into(),
				// on @bluewin.ch
				// - mxbw-bluewin-ch.hdb-cs04.ellb.ch.
				"bluewin-ch.".into(),
				// on @gmx.de, @gmx.ch, @gmx.net
				".gmx.net.".into(),
				// on @icloud.com
				".mail.icloud.com.".into(),
				// on @web.de
				".web.de.".into(),
				".zoho.com.".into(),
			],
		}
	}
}

impl CheckEmailInput {
	/// Create a new CheckEmailInput.
	pub fn new(to_email: String) -> CheckEmailInput {
		CheckEmailInput {
			to_email,
			..Default::default()
		}
	}

	/// Set the email to use in the `MAIL FROM:` SMTP command. Defaults to
	/// `user@example.org` if not explicitly set.
	#[deprecated(since = "0.8.24", note = "Please use set_from_email instead")]
	pub fn from_email(&mut self, email: String) -> &mut CheckEmailInput {
		self.from_email = email;
		self
	}

	/// Set the email to use in the `MAIL FROM:` SMTP command. Defaults to
	/// `user@example.org` if not explicitly set.
	pub fn set_from_email(&mut self, email: String) -> &mut CheckEmailInput {
		self.from_email = email;
		self
	}

	/// Set the name to use in the `EHLO:` SMTP command. Defaults to `localhost`
	/// if not explicitly set.
	#[deprecated(since = "0.8.24", note = "Please use set_hello_name instead")]
	pub fn hello_name(&mut self, name: String) -> &mut CheckEmailInput {
		self.hello_name = name;
		self
	}

	/// Set the name to use in the `EHLO:` SMTP command. Defaults to `localhost`
	/// if not explicitly set.
	pub fn set_hello_name(&mut self, name: String) -> &mut CheckEmailInput {
		self.hello_name = name;
		self
	}

	/// Use the specified SOCK5 proxy to perform email verification.
	#[deprecated(since = "0.8.24", note = "Please use set_proxy instead")]
	pub fn proxy(&mut self, proxy_host: String, proxy_port: u16) -> &mut CheckEmailInput {
		self.proxy = Some(CheckEmailInputProxy {
			host: proxy_host,
			port: proxy_port,
			..Default::default()
		});
		self
	}

	/// Use the specified SOCK5 proxy to perform email verification.
	pub fn set_proxy(&mut self, proxy: CheckEmailInputProxy) -> &mut CheckEmailInput {
		self.proxy = Some(proxy);
		self
	}

	/// Set the number of SMTP retries to do.
	pub fn set_retries(&mut self, retries: usize) -> &mut CheckEmailInput {
		self.retries = retries;
		self
	}

	/// Add optional timeout for the SMTP verification step.
	#[deprecated(since = "0.8.24", note = "Please use set_smtp_timeout instead")]
	pub fn smtp_timeout(&mut self, duration: Duration) -> &mut CheckEmailInput {
		self.smtp_timeout = Some(duration);
		self
	}

	/// Change the SMTP port.
	pub fn set_smtp_port(&mut self, port: u16) -> &mut CheckEmailInput {
		self.smtp_port = port;
		self
	}

	/// Set the SMTP client security to use for TLS.
	pub fn set_smtp_security(&mut self, smtp_security: SmtpSecurity) -> &mut CheckEmailInput {
		self.smtp_security = smtp_security;
		self
	}

	/// Add optional timeout for the SMTP verification step. This is the
	/// timeout for _each_ SMTP connection attempt, not for the whole email
	/// verification process.
	pub fn set_smtp_timeout(&mut self, duration: Option<Duration>) -> &mut CheckEmailInput {
		self.smtp_timeout = duration;
		self
	}

	/// Set whether to use Yahoo's API or connecting directly to their SMTP
	/// servers. Defaults to true.
	#[deprecated(since = "0.8.24", note = "Please use set_yahoo_use_api instead")]
	pub fn yahoo_use_api(&mut self, use_api: bool) -> &mut CheckEmailInput {
		self.yahoo_use_api = use_api;
		self
	}

	/// Set whether to use Yahoo's API or connecting directly to their SMTP
	/// servers. Defaults to true.
	pub fn set_yahoo_use_api(&mut self, use_api: bool) -> &mut CheckEmailInput {
		self.yahoo_use_api = use_api;
		self
	}

	/// Set whether to use Gmail's API or connecting directly to their SMTP
	/// servers. Defaults to false.
	pub fn set_gmail_use_api(&mut self, use_api: bool) -> &mut CheckEmailInput {
		self.gmail_use_api = use_api;
		self
	}

	/// Set whether to use Microsoft 365's OneDrive API or connecting directly
	/// to their SMTP servers. Defaults to false.
	pub fn set_microsoft365_use_api(&mut self, use_api: bool) -> &mut CheckEmailInput {
		self.microsoft365_use_api = use_api;
		self
	}

	/// Whether to check if a gravatar image is existing for the given email.
	/// Defaults to false.
	pub fn set_check_gravatar(&mut self, check_gravatar: bool) -> &mut CheckEmailInput {
		self.check_gravatar = check_gravatar;
		self
	}

	/// Whether to haveibeenpwned' API for the given email
	/// check only if the api_key is set
	pub fn set_haveibeenpwned_api_key(&mut self, api_key: Option<String>) -> &mut CheckEmailInput {
		self.haveibeenpwned_api_key = api_key;
		self
	}

	/// Set whether or not to use a headless navigator to navigate to Hotmail's
	/// password recovery page to check if an email exists. If set to
	/// `Some(<endpoint>)`, this endpoint must point to a WebDriver process,
	/// usually listening on http://localhost:4444. Defaults to None.
	#[cfg(feature = "headless")]
	pub fn set_hotmail_use_headless(
		&mut self,
		use_headless: Option<String>,
	) -> &mut CheckEmailInput {
		self.hotmail_use_headless = use_headless;
		self
	}

	/// **IMPORTANT:** This is a beta feature, and might be completely removed,
	/// or moved somewhere else, before the next release.
	///
	/// List of domains to skip when doing an SMTP connection, because we know
	/// they return "unknown".
	pub fn set_skipped_domains(&mut self, domains: Vec<String>) -> &mut CheckEmailInput {
		self.skipped_domains = domains;
		self
	}
}

/// An enum to describe how confident we are that the recipient address is
/// real.
#[derive(Debug, Eq, PartialEq, Deserialize, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum Reachable {
	/// The email is safe to send.
	Safe,
	/// The email address appears to exist, but has quality issues that may
	/// result in low engagement or a bounce. Emails are classified as risky
	/// when one of the following happens:
	/// - catch-all email,
	/// - disposable email,
	/// - role-based address,
	/// - full inbox.
	Risky,
	/// Emails that don't exist or are syntactically incorrect. Do not send to
	/// these emails.
	Invalid,
	/// We're unable to get a valid response from the recipient's email server.
	Unknown,
}

/// The result of the [check_email](check_email) function.
#[derive(Debug)]
pub struct CheckEmailOutput {
	/// Input by the user.
	pub input: String,
	pub is_reachable: Reachable,
	/// Misc details about the email address.
	pub misc: Result<MiscDetails, MiscError>,
	/// Details about the MX host.
	pub mx: Result<MxDetails, MxError>,
	/// Details about the SMTP responses of the email.
	pub smtp: Result<SmtpDetails, SmtpError>,
	/// Details about the email address.
	pub syntax: SyntaxDetails,
}

impl Default for CheckEmailOutput {
	fn default() -> Self {
		CheckEmailOutput {
			input: String::default(),
			is_reachable: Reachable::Unknown,
			misc: Ok(MiscDetails::default()),
			mx: Ok(MxDetails::default()),
			smtp: Ok(SmtpDetails::default()),
			syntax: SyntaxDetails::default(),
		}
	}
}

// Implement a custom serialize.
impl Serialize for CheckEmailOutput {
	fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
	where
		S: Serializer,
	{
		// This is just used internally to get the nested error field.
		#[derive(Serialize)]
		struct MyError<E> {
			error: E,
			// We add an optional "description" field when relevant, given by
			// the `get_description` on SmtpError.
			#[serde(skip_serializing_if = "Option::is_none")]
			description: Option<SmtpErrorDesc>,
		}

		let mut map = serializer.serialize_map(Some(1))?;
		map.serialize_entry("input", &self.input)?;
		map.serialize_entry("is_reachable", &self.is_reachable)?;
		match &self.misc {
			Ok(t) => map.serialize_entry("misc", &t)?,
			Err(error) => map.serialize_entry(
				"misc",
				&MyError {
					error,
					description: None,
				},
			)?,
		}
		match &self.mx {
			Ok(t) => map.serialize_entry("mx", &t)?,
			Err(error) => map.serialize_entry(
				"mx",
				&MyError {
					error,
					description: None,
				},
			)?,
		}
		match &self.smtp {
			Ok(t) => map.serialize_entry("smtp", &t)?,
			Err(error) => map.serialize_entry(
				"smtp",
				&MyError {
					error,
					description: error.get_description(),
				},
			)?,
		}
		map.serialize_entry("syntax", &self.syntax)?;
		map.end()
	}
}

#[cfg(test)]
mod tests {
	use super::CheckEmailOutput;
	use async_smtp::smtp::response::{Category, Code, Detail, Response, Severity};

	#[test]
	fn should_serialize_correctly() {
		// create a dummy CheckEmailOutput, with a given message as a transient
		// SMTP error.
		fn dummy_response_with_message(m: &str) -> CheckEmailOutput {
			let r = Response::new(
				Code {
					severity: Severity::TransientNegativeCompletion,
					category: Category::MailSystem,
					detail: Detail::Zero,
				},
				vec![m.to_string(), "8BITMIME".to_string(), "SIZE 42".to_string()],
			);

			CheckEmailOutput {
				input: "foo".to_string(),
				is_reachable: super::Reachable::Unknown,
				misc: Ok(super::MiscDetails::default()),
				mx: Ok(super::MxDetails::default()),
				syntax: super::SyntaxDetails::default(),
				smtp: Err(super::SmtpError::SmtpError(r.into())),
			}
		}

		let res = dummy_response_with_message("blacklist");
		let actual = serde_json::to_string(&res).unwrap();
		// Make sure the `description` is present with IpBlacklisted.
		let expected = r#"{"input":"foo","is_reachable":"unknown","misc":{"is_disposable":false,"is_role_account":false,"gravatar_url":null,"haveibeenpwned":null},"mx":{"accepts_mail":false,"records":[]},"smtp":{"error":{"type":"SmtpError","message":"transient: blacklist"},"description":"IpBlacklisted"},"syntax":{"address":null,"domain":"","is_valid_syntax":false,"username":"","normalized_email":null,"suggestion":null}}"#;
		assert_eq!(expected, actual);

		let res =
			dummy_response_with_message("Client host rejected: cannot find your reverse hostname");
		let actual = serde_json::to_string(&res).unwrap();
		// Make sure the `description` is present with NeedsRDNs.
		let expected = r#"{"input":"foo","is_reachable":"unknown","misc":{"is_disposable":false,"is_role_account":false,"gravatar_url":null,"haveibeenpwned":null},"mx":{"accepts_mail":false,"records":[]},"smtp":{"error":{"type":"SmtpError","message":"transient: Client host rejected: cannot find your reverse hostname"},"description":"NeedsRDNS"},"syntax":{"address":null,"domain":"","is_valid_syntax":false,"username":"","normalized_email":null,"suggestion":null}}"#;
		assert_eq!(expected, actual);

		let res = dummy_response_with_message("foobar");
		let actual = serde_json::to_string(&res).unwrap();
		// Make sure the `description` is NOT present.
		let expected = r#"{"input":"foo","is_reachable":"unknown","misc":{"is_disposable":false,"is_role_account":false,"gravatar_url":null,"haveibeenpwned":null},"mx":{"accepts_mail":false,"records":[]},"smtp":{"error":{"type":"SmtpError","message":"transient: foobar"}},"syntax":{"address":null,"domain":"","is_valid_syntax":false,"username":"","normalized_email":null,"suggestion":null}}"#;
		assert_eq!(expected, actual);
	}
}
